tests/5.1/teams/test_target_get_max_teams.c

```diff
@@ -0,0 +1,56 @@
+//===--------------------- test_target_get_max_teams.c ----------------------===//
+//
+// OpenMP API Version 5.1 Nov 2020
+//
+// This test uses the omp_get_max_teams routine to check what the max teams 
+// capacity is for this device. It should return the total amount of teams
+// that could be allocated to a specific teams region
+//
+//===------------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1024
+
+int main() {
+
+        OMPVV_TEST_OFFLOADING;
+
+        int errors = 0;
+        int MAX_TEAMS = omp_get_max_teams(); //on host
+        int num_teams = MAX_TEAMS + 1; //a value that is not possible
+        printf("max on host:%d\n",MAX_TEAMS);
+        #pragma omp teams
+        {
+                if (omp_get_team_num() == 0) {
+                        num_teams = omp_get_num_teams();
+                }
+        }       
+
+        OMPVV_ERROR_IF(MAX_TEAMS > 0 && num_teams > MAX_TEAMS, "Number of teams reported on host exceeded max number of teams (max no. > 0)");
+        OMPVV_ERROR_IF(num_teams <= 0, "Number of teams reported on host is 0 or negative");
+        OMPVV_TEST_AND_SET(errors, num_teams > MAX_TEAMS);
```

Comment:
Actually, I feel the other way around, that is MAX_TEAMS > 0 should be removed. As per the 5.1/5.2 spec omp_get_max_teams gets the value from the nteams-var ICV of the relevant device. But the fact that the thread(s) enter the teams region means that there is atleast 1 team. But the ICV can still be set to zero. Which is what I am seeing with gcc right now.

Reply:
As you said, the ICV can remain as zero unless either a user or the runtime sets it differently, and that's why comparing `num_teams > MAX_TEAMS` is meaningful only if the ICV is set correctly (`MAX_TEAMS > 0`).

Reply:
I have an idea for this test. Given that the initial value of `nteams-var` is set by the `OMP_NUM_TEAMS` variable, and what we testing is `omp_get_max_teams()` which return this value. And given that the initial value of `nteams-var` is zero (I understand that this is when it is not set by the environment variable, which I think it makes no sense). We could test this, by setting the environment variable to say `2`, and testing that that value is read by the function. This way, I think it will pass even if in the future the specification changes the default value of the `nteams-var`.

Reply:
I agree that num_teams > MAX_TEAMS is meaningful if MAX_TEAMS > 0 is also true. Otherwise, the behavior is implementation defined, per section 6.23 of the 5.1 spec. However, per that same section, "The behavior of the program is implementation defined if the requested value of OMP_NUM_TEAMS is greater than the number of teams that an implementation can support..." As such I agree that there should be set and get tests for the environmental variable (done manually and with the omp_set_num_teams() routine) in order to check the resulting behavior for various values. From the results of this test, it seems true that, unmodified, the initial value of the ICV is 0 per device. That case I agree doesn't make sense, but gcc has some documentation on its expected implentation defined behavior in that case https://github.com/gcc-mirror/gcc/blob/master/libgomp/testsuite/libgomp.c-c%2B%2B-common/teams-nteams-icv-1.c
GCC developers also include set and get tests in that file, along with tests for setting the env variable in https://github.com/gcc-mirror/gcc/blob/master/libgomp/testsuite/libgomp.c-c%2B%2B-common/teams-nteams-icv-2.c and other related files. LLVM developers have variations of set and get tests:
https://github.com/llvm/llvm-project/blob/main/openmp/runtime/test/api/omp_nteams_api.c
https://github.com/llvm/llvm-project/blob/main/openmp/runtime/test/api/omp_nteams_api_env.c
https://github.com/llvm/llvm-project/blob/main/openmp/runtime/test/api/omp_nteams_api_env.2c